src/drivers/sdr/gnuradio.c

```diff
@@ -0,0 +1,206 @@
+            mtu = fec_get_next_mpdu(ifdata->mac_data, (void **)&buf);
+            csp_sleep_ms(delay_time);
+        }
+    }
+    return CSP_ERR_NONE;
+}
+
+CSP_DEFINE_TASK(csp_gnuradio_rx_task) {
+
+    usart_context_t * ctx = arg;
+
+    uint8_t rxframe[128] = {0};
+    int len = 128;
+
+    // Receive loop
+    while (1){
+        rc = zmq_recv(subscriber, rxframe, len, 0);//will this wait here until data is available?
+        //^^^verify that gnuradio passes frame correctly and sync word and length bytes ignored
+
+        assert(rc != -1);
+
+        ctx->rx_callback(ctx->user_data, rxframe, len, NULL);
+    }
+    return NULL;
+}
+
+//format and send 128B frame to gnuradio via zmq
+static void csp_gnuradio_tx (int fd, const void * data, size_t len){
+    if(len != 128){
+        while(1); //improve error handling
+    }
+
+    //apply framing according to UHF user manual protocol
+    uint8_t crc_command[128] = {0};
+    crc_command[0] = len;
+    for(int i = 0; i < len; i++) {
+        crc_command[1+i] = data[i];
+    }
+
+    uint16_t crc_res = crc16(crc_command, len+1);
+
+    uint8_t radio_command[148] = {0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA,
+                                  0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA}; // Gives 16 preamble bytes
+    radio_command[16] = 0x7E;
+    radio_command[17] = len;
+    for( int i = 0; i < len; i++) {
+        radio_command[18+i] = data[i];
+    }
+
+    radio_command[18+len] = ((uint16_t)crc_res >> 8) & 0xFF;
+    radio_command[18+len+1] = ((uint16_t)crc_res >> 0) & 0xFF;
+
+    //send to gnuradio via zmq
+    int radio_len = 16+2+len+2;//preamble + sync word + length indicator + data + crc
+
+    rc = zmq_send(publisher, radio_command, radio_len, 0);
+    assert(rc == radio_len);
+
+    return CSP_ERR_NONE;
+}
+
+void csp_sdr_rx(void *cb_data, uint8_t *buf, size_t len, void *pxTaskWoken) {
+    if(len != 128){
+        while(1);//improve error handling
+    }
+    
+    csp_iface_t *iface = (csp_iface_t *) cb_data;
+    csp_sdr_interface_data_t *ifdata = (csp_sdr_interface_data_t *)iface->interface_data;
+    uint8_t *data = (uint8_t *)buf;
+    while (len--) {
+        if (csp_queue_enqueue(ifdata->rx_queue, (const uint8_t *)data, QUEUE_NO_WAIT) != true) {
+            return;
+        }
+        data++;
+    }
+}
+
+//Inits zmq tx(pub) and rx(sub)
+//starts rx thread
+int csp_gnuradio_open(const csp_usart_conf_t *conf, csp_usart_callback_t rx_callback, void * user_data,  csp_usart_fd_t* return_fd) {
+
+    context = zmq_ctx_new();
+
+    //Init zmq tx(pub)
+    publisher = zmq_socket(context, ZMQ_PUB);
+    int rc = zmq_bind(publisher, "tcp://127.0.0.1:0001");
+    assert(rc == 0);
+
+    // and rx(sub)
+    subscriber = zmq_socket(context, ZMQ_SUB);
+    int rc = zmq_connect(subscriber, "tcp://127.0.0.1:0002");
+    assert(rc == 0);
+    rc = zmq_setsockopt(subscriber, ZMQ_SUBSCRIBE, "", 0);
+    assert(rc == 0);
+
+    usart_context_t * ctx = calloc(1, sizeof(*ctx));
```

Comment:
Better to do sizeof(usart_context_t). Also, you must cast the return value of calloc to usart_context_t *

I'm also wondering what the purpose of ctx is